### DISCUSSION TITLE
How to get the current connector chainId? How do I detect a network change?

### DISCUSSION THREAD

**Discussion by pepoospina** [wevm/wagmi#972] [Q&A] [answered]
It seems that `useChainId()` is not really publicly supported, so I'm getting the chainId of the currently connected chain (on Metamask) from 

```
const [chainId, setChainId] = useState<number>();

const { connectAsync } = useConnect({
  connector: new InjectedConnector(),
});

const connectResult = await connectAsync();
setChainId(connectResult.chain.id);
```

But now I can't figure out how can I react to the user changing the network on metamask. I would need to update chainId, but I can't see any `onNetworkChange()` function anywhere...

**Comment by tmm** [comment, marked as answer]
Check out [`useNetwork`](https://wagmi.sh/docs/hooks/useNetwork):

```tsx
import { useNetwork } from 'wagmi'

function App() {
  const { chain } = useNetwork()
}
```

**Comment by Evirtual** [comment]
How would I get current chain ID with Wagmi version 2.1.2? I used `const chainId = useChainId();`, but it always return "1". I used WalletConnect option to connect.

**Reply by tmm** [reply]
The result of `useChainId` should change when the connector switches chains. You can also use [`useAccount#chainId`](https://wagmi.sh/react/api/hooks/useAccount#chainid) for the connected `chainId`.

**Reply by Evirtual** [reply]
Hi. Thank you for reply, I tried using `chainId` from `useAccount`, but it always returns the first chainId from wagmi config list, even though I switch networks in a wallet. We are currently testing WalletConnect -> Internet Money Wallet. here is my config for reference:

```javascript
import { http, createConfig } from 'wagmi'
import { pulsechain, pulsechainV4 } from 'wagmi/chains'
import { walletConnect } from 'wagmi/connectors'

const projectId = process.env.REACT_APP_WALLETCONNECT_PROJECT_ID || '';

export const config = createConfig({
  chains: [pulsechain, pulsechainV4],
  connectors: [
    walletConnect({ 
      projectId
    }),
  ],
  transports: {
    [pulsechain.id]: http(),
    [pulsechainV4.id]: http(),
  },
})
```

**NOTE**: If I use MetaMask to connect it works. So my guess it is related to WalletConnect v2 provider. Since this `await connection.connector.getProvider({ chainId })` fails with error `getProvider(); is not a function` when I connect with WalletConnect option using wagmi setup above.

At the moment I applied a temporary solution: `const chainId = process.env.REACT_APP_ENVIRONMENT === 'prod' 
 REACT_APP_CHAIN_ID : REACT_APP_CHAIN_ID_TESTNET ;`, since we only use one network, but it is not optimal for multiple chain setup.

**Reply by einaralex** [reply]
I'm struggling with this as well; I don't know at what point it stopped working; we are at the latest Wagmi v1.
When connecting a Safe through WalletConnect, it is used to feed the Safes connected chain through and connect to that chain.
Now, there seems to be no way to get the chain from the wallet; always just defaults to the first one in the list that is initialized.

**Reply by tmm** [reply]
Please share a minimal reproduction.

**Reply by einaralex** [reply]
Here is an example of our setup: [CODESANDBOX](https://codesandbox.io/p/sandbox/rainbowkit-sign-3sth7y?file=%2Fsrc%2FApp.js%3A16%2C40&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu95920d00063b6u9rnc41nk%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu95920c00023b6uly5uw3ch%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu95920c00033b6uay6hvuko%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu95920c00053b6uelufjjuh%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B40%252C60%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu95920c00023b6uly5uw3ch%2522%253A%257B%2522id%2522%253A%2522clu95920c00023b6uly5uw3ch%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu96g0le00023b6um53w7235%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FWalletProvider.js%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A48%252C%2522startColumn%2522%253A31%252C%2522endLineNumber%2522%253A48%252C%2522endColumn%2522%253A31%257D%255D%257D%252C%257B%2522id%2522%253A%2522clu97ich600023b6u8o9614t9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A12%252C%2522startColumn%2522%253A22%252C%2522endLineNumber%2522%253A12%252C%2522endColumn%2522%253A22%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cluk1v64b00023b6mkayxnjtr%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A16%252C%2522startColumn%2522%253A40%252C%2522endLineNumber%2522%253A16%252C%2522endColumn%2522%253A40%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cluk1v64b00023b6mkayxnjtr%2522%257D%252C%2522clu95920c00053b6uelufjjuh%2522%253A%257B%2522id%2522%253A%2522clu95920c00053b6uelufjjuh%2522%252C%2522activeTabId%2522%253A%2522clu95m1br006e3b6uo7dmc0ww%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522id%2522%253A%2522clu95m1br006e3b6uo7dmc0ww%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%252C%257B%2522type%2522%253A%2522ASSISTANT_CHAT%2522%252C%2522chatId%2522%253A%2522clu99dfiq011s3b6ut5xv6cxa%2522%252C%2522id%2522%253A%2522clu99dfis011u3b6ucer6l2jc%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clu95920c00033b6uay6hvuko%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clu95920c00033b6uay6hvuko%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15.403551251008892%257D)

I try connecting with a Sepolia wallet, but it will always default to Gnosis Chiado because it's the first one in the chain list.
As I understood, useNetwork should show the wallets chain, but this example shows that it's not working correctly.

1. Create Sepolia Safe on https://app.safe.global/
2. Go to Home > Use WalletConnect and pair the Safe to the Codesandbox.
3. Observer that chain from useNetwork is Gnosis Chiado, not Sepolia.

**Reply by tmm** [reply]
Got it. Wagmi is onto v2 now so v1 is only updated for critical security issues.

**Reply by einaralex** [reply]
Totally understand that, but the problem is still there in v2. It can be tested on the rainbowkit home page, it defaults to Ethereum instead of the network for the Safe

https://www.rainbowkit.com/

**Reply by marcusmolchany** [reply]
I was having the same issue with `useChainId()`. When I changed networks in MetaMask it was not updating. I switched to using `const { chainId} = useAccount()` and that works as expected.

**Reply by tanlm1996** [reply]
This works. Thank you very much!
